lib/runner/variant-reassignment.js

```diff
@@ -212,14 +216,96 @@ export default class VariantReassignment {
+   *
+   * @param productDraft
+   * @param matchingProducts
+   * @param ctpProductToUpdate
+   * @returns {{matchingProducts, ctpProductToUpdate}}
+   * @private
+   */
+  _getRemovedVariants (productDraft, matchingProducts, ctpProductToUpdate) {
+    const productsToRemoveVariants = matchingProducts.filter(p => p !== ctpProductToUpdate)
+    const skus = this._getProductDraftSkus(productDraft)
+    const collectVariantsWithCondition = (product, condition) => {
+      const currentVariants = this._getProductVariants(product, true)
+      const stagedVariants = this._getProductVariants(product)
+
+      const variants = stagedVariants
+
+      // if variant is only in current and not in staged, push it to variants collection
+      currentVariants.forEach((current) => {
+        const isInStaged = stagedVariants.some(staged => staged.sku === current.sku)
+        if (!isInStaged)
+          variants.push(current)
+      })
+
+      const variantsToBeCollected = variants.filter(condition)
+      if (variantsToBeCollected.length)
+        return variantsToBeCollected
+      return []
+    }
+
+    // variants that needs to be moved from matching product
+    const matchingProductsVariants = productsToRemoveVariants.map(product =>
+      collectVariantsWithCondition(product, variant => skus.includes(variant.sku))
+    )
+    // variants that needs to be removed from CTP product to update
+    const ctpProductToUpdateVariants
+      = collectVariantsWithCondition(ctpProductToUpdate, variant => !skus.includes(variant.sku))
+
+    return {
+      matchingProductsVars: matchingProductsVariants,
+      ctpProductToUpdateVars: ctpProductToUpdateVariants
+    }
+  }
+
+  _createProductDraftWithRemovedVariants (product, variantsToBackup) {
+    let productDraftClone
+    if (variantsToBackup.length > 0) {
+      productDraftClone = _.cloneDeep(product.masterData.staged)
+      productDraftClone.key = product.key
+      productDraftClone.productType = product.productType
+      productDraftClone.taxCategory = product.taxCategory
+      productDraftClone.state = product.state
+      productDraftClone.reviewRatingStatistics = product.reviewRatingStatistics
+      productDraftClone.masterVariant = variantsToBackup[0]
+      productDraftClone.variants = variantsToBackup.slice(1, variantsToBackup.length)
+    }
+
+    return productDraftClone
+  }
+
+  _backupToCustomObject (newProductDraft, backupProductDraft, variants) {
+    return this.transactionService.createTransaction({
+      newProductDraft,
+      backupProductDraft,
+      variants
+    })
+  }
+
   _fetchMatchingProductsForReassignment (productDraft) {
-    const skus = this._getProductSkus(productDraft)
+    const skus = this._getProductDraftSkus(productDraft)
     return this.productService.getProductsBySkus(skus)
   }
 
-  _getProductSkus (product) {
-    return [product.masterVariant.sku]
-      .concat(product.variants.map(v => v.sku))
+  _getProductDraftSkus (product) {
```

Comment:
This can be simplified to this:
```js
 _getProductDraftSkus (product) {
    const variants = [product.masterVariant].concat(product.variants || [])
    return _.map(variants, 'sku')
  }
```

Reply:
https://github.com/commercetools/commercetools-node-variant-reassignment/pull/23/commits/e4cd1fc5c7b1b9bcc9b3ee41dd90f9e1e2d711c8#diff-15ce2e3c2d77d574a8c8ae56d19b9d92R274